### PR TITLE
Initial FN Unifier Pull Request

### LINKED ISSUE #238: [code2fn] code2fn unifier for enriching function networks with extracted comments
Provide a function that takes a) comment extraction output and b) a function network (GrometModuleCollection) and returns a function network (GrometModuleCollection) enriched with comment metadata.

### PR DESCRIPTION
**Pull request by titomeister**
This pull request introduces a new program that is used to unify a GroMEt FN and its associated comment extractions into a GroMEt FN with comments aligned to it.
This is still in its early stages, as it doesn't fully handle all the alignment cases in which comments and docstrings appear, including but not limited to:
- Comments that directly reference variables
- Docstrings that are not directly a part of a function definition.

Resolves #238

### CODE REVIEW

**Comment by myedibleenso** [comment]
@titomeister , assuming I have a) gromet JSON (python dict) and b) extracted comments (python dict), what function should I be calling to combine the two into a new gromet JSON?

**Comment by titomeister** [comment]
> @titomeister , assuming I have a) gromet JSON (python dict) and b) extracted comments (python dict), what function should I be calling to combine the two into a new gromet JSON?

You can call the script from the command line as follows:
`python fn_unifier --gromet <GROMET_JSON> --comments <COMMENTS_JSON>`

There's currently no independent function that you can call to combine the two, you have to run the unifier from main.
However I was working on a function to do just that so you don't have to run it from the command line. I just didn't get it done in time for the PR.

**Comment by myedibleenso** [comment]
>I was working on a function to do just that so you don't have to run it from the command line. I just didn't get it done in time for the PR.

In that case, let's wait to merge this until that function is ready.  We'll need it for the REST-based calls.

**Comment by adarshp** [comment]
Note: We may not necessarily need this function, depending on when the tree-sitter based comment extraction is implemented...

**Comment by titomeister** [comment]
I went ahead and wrote a small one regardless of whether or not we need it. I wanted a way to expose the interface such that another program could use it without having to resort to calling the entire program on its own.
I'm working out some small details still, but will update soon.

**Comment by myedibleenso** [comment]
@adarshp yesterday the PA team discussed a tentative plan and estimated development costs for porting comment extraction to tree-sitter.  Based on that discussion, we would continue to use the unifier.

**Comment by adarshp** [comment]
Sounds good, thanks for the update @myedibleenso !

File: skema/program_analysis/fn_unifier.py [hunk 1]
```diff
@@ -188,20 +188,20 @@ def align_full_system(gromet_obj, extraction, extraction_file_name):
...
-            # Go through each file in the GroMEt FN
-            normalized_path = normalize_module_path(module)
-            if normalized_path == extraction_file_name:
-                # Find the current FN in the collection
-                module_FN = find_fn(gromet_obj.modules, normalized_path)
-                if module_FN != None:
-                    FN_metadata = module_FN.metadata_collection
-                    align_fn(FN_metadata, extraction, module_FN.fn)
+        if len(gromet_obj.modules[0]) != 1:
+            raise NotImplementedError("Single file alignment from a multi module GroMEt system not supported yet")
```
Comment (myedibleenso) [inline skema/program_analysis/fn_unifier.py:195]: Nice.  This is a good interim solution.

File: skema/program_analysis/fn_unifier.py [hunk 2]
```diff
@@ -221,6 +221,14 @@ def align_full_system(gromet_obj, extraction, extraction_file_name):
                         for FN in module_FN.fn_array:
                             align_fn(FN_metadata, file_comments, FN)
         
+def process_alignment(gromet_json, comments_json):
```
Comment (myedibleenso) [inline skema/program_analysis/fn_unifier.py:224]: @titomeister , can you please add some typehints to this?

Comment (myedibleenso) [reply skema/program_analysis/fn_unifier.py:224]: i.e., (`gromet_json: Dict[Text, Any]`, `comments_json: Dict[Text, Any]` -> GrometFNModuleCollection:`).  Idk if we can be any more specific in the typehint about the structure of `comments_json`.

**Review by myedibleenso — CHANGES_REQUESTED** [review]
Thanks, @titomeister .  Just requesting a couple of minor changes for improved clarity.

File: skema/program_analysis/fn_unifier.py (hunk 2 above)
Comment (titomeister) [reply skema/program_analysis/fn_unifier.py]: Done!

**Review by myedibleenso — APPROVED** [review]

**Pull request merged**